[JSC][GLib] Add support to enable the GLib API when building JSCOnly

spenap opened the pull request:
#### 7bd21a5a9b960ae84aa66df1c8d0f6aa3d989a0c
<pre>
[JSC][GLib] Add support to enable the GLib API when building JSCOnly
<a href="https://bugs.webkit.org/show_bug.cgi?id=296991">https://bugs.webkit.org/show_bug.cgi?id=296991</a>

Reviewed by Adrian Perez de Castro.

This updates the JSCOnly build so that it can optionally include the
GLib API, in the same way used by the WPE and GTK ports.

* Source/JavaScriptCore/GLib.cmake: add FRAMEWORK_HEADERS_DIR and
  DERIVED_SOURCES_DIR like the GTK and WPE ports do.
* Source/JavaScriptCore/PlatformJSCOnly.cmake:
* Source/JavaScriptCore/PlatformJSCOnly.cmake: support building with the
  JSC GLib API by including the GLib cmake, set library output name and
  install PackageConfig file and headers.
* Source/JavaScriptCore/javascriptcoreglib.pc.in: Added.
* Source/WTF/wtf/Platform.h: Guard GLIB versions under USE(GLIB) so they
  also apply when building JSC with the GLib API.
* Source/WTF/wtf/PlatformJSCOnly.cmake: add additionally required
  headers when building with the GLib API.
* Source/cmake/OptionsJSCOnly.cmake: Define ENABLE_JSC_GLIB_API as a
  port option. Set the GLib event loop type and expose required build
  definitions and installation paths.

Canonical link: <a href="https://commits.webkit.org/298565@main">https://commits.webkit.org/298565@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6dd6e0b3d9ce9485a2bc95b54b6fdd787b37e26b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115921 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35575 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26111 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121970 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66451 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36264 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44162 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88057 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118869 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28936 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104027 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68470 "Found 1 new API test failure: /WPE/TestCookieManager:/webkit/WebKitCookieManager/replace-get-all-cookies (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28058 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22135 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65640 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108030 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98323 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22269 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125122 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/114449 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42808 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32129 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96815 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43174 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100215 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96602 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24568 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41861 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19733 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38731 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42697 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48286 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/143133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42164 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/143133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45498 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43871 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/30a9cf57d9389b9ba2e7ecc76e9708ebf1094510. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49004

spenap commented:
This is still a draft -- I am planning to add package config support.

I am sharing this already because I'm curious why I need to use `GLIB_DISABLE_DEPRECATION_WARNINGS` to work around deprecation warnings originating from `Source/WTF/wtf/glib/WTFGType.h` DEFINE_TYPE macros cc/ @aperezdc , @carlosgcampos

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/57d1da4c7fd146d17ed5ba30aabffb02ef15d8d6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49004

carlosgcampos commented on Source/JavaScriptCore/PlatformJSCOnly.cmake:
```diff
@@ -1,3 +1,11 @@
+if (ENABLE_JSC_GLIB_API)
```
I don't know much about cmake, but don't you need to define this option in OptionsJSCOnly.cmake ? We should also make sure that this options depends on enabling the glib event loop

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b339e5f339ebaf91704111f3255676cbf73fd4a2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49004

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/213513d49f68201fde47dda697394c63d458e5d1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49004

carlosgcampos commented on Source/WTF/wtf/Platform.h:
```diff
@@ -82,20 +82,16 @@
 #if USE(GTK4)
 #define GDK_VERSION_MIN_REQUIRED GDK_VERSION_4_0
 #else
 #define GDK_VERSION_MIN_REQUIRED GDK_VERSION_3_22
```
Now we are defining this for other ports, we still need a `#if PLATFORM(GTK)`

spenap replied:
thanks for spotting this!

spenap commented on Source/WTF/wtf/Platform.h:
```diff
@@ -82,20 +82,16 @@
 #define GLIB_VERSION_MAX_ALLOWED GLIB_VERSION_2_70
+#endif
+
 #if USE(GTK4)
```
```suggestion
#if PLATFORM(GTK) && USE(GTK4)
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9baf5ddbbb2331a9c08c8ec8f5d7d2993ef24ea1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49004

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6dd6e0b3d9ce9485a2bc95b54b6fdd787b37e26b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49004

spenap committed:
[JSC][GLib] Add support to enable the GLib API when building JSCOnly
https://bugs.webkit.org/show_bug.cgi?id=296991

Reviewed by Adrian Perez de Castro.

This updates the JSCOnly build so that it can optionally include the
GLib API, in the same way used by the WPE and GTK ports.

* Source/JavaScriptCore/GLib.cmake: add FRAMEWORK_HEADERS_DIR and
  DERIVED_SOURCES_DIR like the GTK and WPE ports do.
* Source/JavaScriptCore/PlatformJSCOnly.cmake:
* Source/JavaScriptCore/PlatformJSCOnly.cmake: support building with the
  JSC GLib API by including the GLib cmake, set library output name and
  install PackageConfig file and headers.
* Source/JavaScriptCore/javascriptcoreglib.pc.in: Added.
* Source/WTF/wtf/Platform.h: Guard GLIB versions under USE(GLIB) so they
  also apply when building JSC with the GLib API.
* Source/WTF/wtf/PlatformJSCOnly.cmake: add additionally required
  headers when building with the GLib API.
* Source/cmake/OptionsJSCOnly.cmake: Define ENABLE_JSC_GLIB_API as a
  port option. Set the GLib event loop type and expose required build
  definitions and installation paths.

Canonical link: https://commits.webkit.org/298565@main
